### DISCUSSION TITLE
Coordinating parallelizable work from a task?

### DISCUSSION THREAD

**Discussion by opsound** [oxidecomputer/hubris#410] [Q&A] [answered]
I read the Hubris [reference](https://oxidecomputer.github.io/hubris/reference/#_sending_messages) and am having trouble envisioning how to achieve certain behavior given the synchronous IPC constraint:

> Sending a message always blocks the sender, waiting for a reply

Suppose task A (orchestrator) needs to perform work which could be run in parallel by tasks B and C (say they each abstract access to some separate HW).

Is there a way to achieve this?  Or an alternative task responsibility pattern to make it possible?

EDIT - reading further, it seems like [notifications](https://oxidecomputer.github.io/hubris/reference/#_what_are_they_good_for) pingback pattern could work.

The task A orchestrator ("client" here) can send to multiple servers (tasks B and C) to kick off the work, and await notifications from each, which represent the completion of said work, and then send again to B and C retrieve the result of the work?

**Comment by steveklabnik** [comment, marked as answer]
Yep, your edit is how I would structure this. The [POST syscall](https://oxidecomputer.github.io/hubris/reference/#_post_11) would let B and C send those notifications to A.
